dot/parachain/bitfield-signing/bitfield_signing.go

```diff
@@ -0,0 +1,226 @@
+	if err != nil {
+		return err
+	}
+	signingContext := parachaintypes.SigningContext{
+		SessionIndex: sessionIndex,
+		ParentHash:   relayParent,
+	}
+	statement := parachaintypes.NewStatementVDT()
+	err = statement.SetValue(bitfield)
+	if err != nil {
+		return err
+	}
+	signature, err := statement.Sign(b.keystore, signingContext, parachaintypes.ValidatorID(validatorID[:]))
+	if err != nil {
+		return err
+	}
+
+	// distribute to subsystem to overseer chan
+	b.subSystemToOverseer <- parachaintypes.DistributeBitfield{
+		RelayParent: relayParent,
+		Bitfield: parachaintypes.UncheckedSignedAvailabilityBitfield{
+			Payload:        bitfield,
+			ValidatorIndex: validatorIndex,
+			Signature:      *signature,
+		},
+	}
+
+	return nil
+}
+
+// ProcessBlockFinalizedSignal processes block finalized signal
+func (b *BitfieldSigning) ProcessBlockFinalizedSignal(parachaintypes.BlockFinalizedSignal) error {
+	return nil
+}
+
+// Stop stops the BitfieldSigning subsystem
+func (b *BitfieldSigning) Stop() {
+	// TODO: anything to do to stop this subsystem ?
+	logger.Infof("Stopping BitfieldSigning subsystem")
+}
+
+func constructAvailabilityBitfield(
+	rt runtime.Instance,
+	validatorIdx parachaintypes.ValidatorIndex,
+	subSystemToOverseer chan<- any,
+) (parachaintypes.BitVec, error) {
+	cores, err := rt.ParachainHostAvailabilityCores()
+	if err != nil {
+		return parachaintypes.BitVec{}, err
+	}
+
+	// init a bitfield without caring the order
+	bitfield := make([]bitfieldData, len(cores))
+
+	var wg sync.WaitGroup
+	for coreOrderIndex, core := range cores {
+		coreValueIndex, value, err := core.IndexValue()
+		if err != nil {
+			return parachaintypes.BitVec{}, err
+		}
+		// 0 is type of parachaintypes.OccupiedCore
+		if coreValueIndex == 0 {
+			wg.Add(1)
+			go func(oi int, v parachaintypes.OccupiedCore) {
+				receivingChan := make(chan bool)
+				queryPayload := availabilitystore.QueryChunkAvailability{
+					CandidateHash:  parachaintypes.CandidateHash{Value: v.CandidateHash},
+					ValidatorIndex: uint32(validatorIdx),
+					Sender:         receivingChan,
+				}
+				// send QueryChunkAvailability to availability store via overseer
+				subSystemToOverseer <- queryPayload
+
+				// append the result to the bitfield
+				bitfield = append(bitfield, bitfieldData{index: oi, data: <-receivingChan})
+			}(coreOrderIndex, value.(parachaintypes.OccupiedCore))
+		} else {
+			bitfield = append(bitfield, bitfieldData{index: coreOrderIndex, data: false})
+		}
+	}
+	wg.Wait()
+
+	return bitfieldOrderGuard(bitfield), nil
+}
+
+// bitfieldOrderGuard sort the fulfilled bitfield data in its correct order according to the CoreState
+func bitfieldOrderGuard(bitfieldData []bitfieldData) parachaintypes.BitVec {
+	sort.Slice(bitfieldData, func(i, j int) bool {
+		return bitfieldData[i].index < bitfieldData[j].index
+	})
+
+	b := make([]bool, len(bitfieldData))
+	for _, data := range b {
+		b = append(b, data)
+	}
```

Comment:
question. this iteration over the created slice `b` is needed?

Reply:
lol man, good catch, pushing fix now